Echosa's Dota2 Spice Instant Answer

echosa opened the pull request:
# Spice Pull Request Template

Using this template will help us better understand your Instant Answer and assist you when necessary.  Simply copy and paste the markdown below into the description of your GitHub pull request and complete as appropriate.

**What does your Instant Answer do?**
This instant answers give quick access to information about Dota 2 heroes and items.

**What problem does your Instant Answer solve (Why is it better than organic links)?**
The instant answer makes it faster to get the information versus having to navigate pages. Also, it makes DDG just that much more desirable for users, both current and potential.

**What is the data source for your Instant Answer? (Provide a link if possible)**
http://www.dota2.com/heroes/
http://www.dota2.com/items/
https://www.dota2.com/jsfeed/heropediadata?feeds=itemdata,herodata,abilitydata

**Why did you choose this data source?**
It is the official Dota2 source.

**Are there any other alternative (better) data sources?**
Not that I found.

**What are some example queries that trigger this Instant Answer?**
dota2 blink
dota 2 pudge
crystal maiden dota2
clarity dota 2
dota2 invoker
dota2 magic wand

**Which communities will this Instant Answer be especially useful for? (gamers, book lovers, etc)**
gamers, dota 2 players and enthusiasts

**Is this Instant Answer connected to a DuckDuckHack [Instant Answer idea](https://duck.co/ideas)?**
No. I didn't find any ideas for a Dota2 IA.

**Which existing Instant Answers will this one supersede/overlap with?**
N/A

**Are you having any problems? Do you need our help with anything?**
~~Cross-browser testing. It works for me fine for me in Firefox (OS X), but not Chrome for some reason.~~
This is no longer an issue.

**What are the terms of use for the API? Will DuckDuckGo need specific authorization (e.g. an API key)? Are there any costs associated with API usage?**
It's just a json endpoint from dota2.com. No keys or auth necessary.

**What does the Instant Answer look like? (Provide a screenshot for new or updated Instant Answers)**
![screen shot 2015-06-27 at 11 25 31 am](https://cloud.githubusercontent.com/assets/133786/8392968/e9119d80-1cc0-11e5-91a1-736eb737e77e.png)
![screen shot 2015-06-27 at 11 25 57 am](https://cloud.githubusercontent.com/assets/133786/8392970/e924aae2-1cc0-11e5-9f01-a39c06110a2d.png)
![screen shot 2015-06-27 at 11 26 12 am](https://cloud.githubusercontent.com/assets/133786/8392969/e9226cdc-1cc0-11e5-9675-5513e8756f68.png)


## Checklist
Please place an 'X' where appropriate.

```
[X] Added metadata and attribution information
[X] Wrote test file and added to t/ directory
[X] Verified that instant answer adheres to design guidelines (https://duck.co/duckduckhack/design_styleguide)
[X] Verified that instant answer adheres to code styleguide (https://duck.co/duckduckhack/code_styleguide)
[X] Tested cross-browser compatibility

    Please let us know which browsers/devices you've tested on:
    - Windows 8
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 10

    - Windows 7
        [] Google Chrome
        [] Firefox
        [] Opera
        [] IE 8
        [] IE 9
        [] IE 10

    - Windows XP
        [] IE 8

    - Mac OSX
        [X] Google Chrome
        [X] Firefox
        [] Opera
        [X] Safari

    - iOS (iPhone)
        [X] Safari Mobile
        [] Google Chrome
        [] Opera

    - iOS (iPad)
        [] Safari Mobile
        [] Google Chrome
        [] Opera

    - Android
        [] Firefox
        [] Native Browser
        [] Google Chrome
        [] Opera
```
-----
IA Page: https://duck.co/ia/view/dota2

moollaza commented:
@echosa thanks a lot for contributing, we really appreciate it!

It's clear that you spent a lot of time on this and we're grateful, but unfortunately I can't accept this in it's current state. We ask in our documentation that you email us for approval so we can verify the source and help you plan the IA, with the goal being that you won't end up creating a fully custom HTML template like you've used here and also that the code is able to remain simple (ideally no HTML or CSS needed). We want to make sure that the IA's maintain visual consistency and styling and I feel that this IA currently has a little too much going on.

I hate to ask, but I need you to please take a step back and think about how which of this information is really important and take a look at our templates and examples to see which templates work best. I believe our **Info** template (with the Infobox) should work perfectly for this kind of IA and there are several live examples of very similar Instant Answers (e.g. Magic the Gathering, Pokemon -- checkout https://duck.co/ia/). Some of what you're showing will need to be sacrificed though -- we can't show *all* of this, and that's why the More at link exists, to send users to see things in greater detail.

------

Separately, we're really curious to know what your process was in developing this. Did you read through the docs? Were they easy to follow? Please send me an email (moollaza@duckduckgo.com) as we'd love to speak with you!

moollaza commented:
> It's just a json endpoint from dota2.com. No keys or auth necessary.

@echosa are there no other API's out there? I'm very reluctant to use this as an "API" because the size of the response is huge -- 700Kb!

echosa commented:
> We ask in our documentation that you email us for approval so we can verify the source and help you plan the IA

I don't recall seeing that. I must have missed it. Sorry!

> We want to make sure that the IA's maintain visual consistency and styling

I agree with this, and I want to match the expected style as well. The docs didn't help much, so I went ahead and wrote what I've submitted, figuring someone would call me out and help get it right. Getting the backend working and passing variables into the template was more the goal. The templates can be rearranged and redesigned as needed and desired. 

>  I believe our Info template (with the Infobox) should work perfectly for this kind of IA

Well, at least I was partially on the right path, since I also determined "info" would be the right choice, and made that the group that is used. As for the templates, I wasn't quite sure what to do. The documentation didn't seem clear to me at the time I developed this (which was a few hours each on Thursday and Friday, mostly).

> Some of what you're showing will need to be sacrificed though

Sad. I was really hoping for a comprehensive answer, so having to go to the source wasn't necessary. However, if I do have to compromise and sacrifice some data, I'm sure I can come up with something.

> Separately, [...] Did you read through the docs? Were they easy to follow? Please send me an email [...]

I will send an email, but the short answers to both of those questions are "Yes" and "Mostly, but not always".

> are there no other API's out there? [...] the response is huge

I didn't find any other APIs, and this seemed the best as it is the actual source, not a third-party site. I was also concerned about the size, and was wonder if there's some kind of caching that can be done to keep a local copy of the API response, that can be checked and invalidated in/when the API response changes. (Do y'all have cron jobs that can run to do something like this? That's probably overstepping the bounds of what an IA can do.)

moollaza commented:
Sorry for the delay @echosa!

@abeyang I spent some time playing around with our templates to see how far I could push them and realized we can pass along our existing icon template, as a sub-template to the List templ -- meaning we have a list of icon-template results.

@echosa has taken my code and improved upon it and we're wondering what you think of the current state. My goal was to try and use the List so the initial display isn't too overwhelming, but still allows us to display a lot useful information.

Any thoughts?

abeyang commented:
@echosa , forgive me for my complete lack of knowledge of Dota2 (gasp), but this is what I'm seeing on my dev machine as of yesterday:

![image](https://cloud.githubusercontent.com/assets/794084/9068033/bfaa0282-3a95-11e5-8997-0d4e9b40cea6.png)

Few questions:
- Is that the largest thumbnail available to us? Can we get a bigger one?
- What do the six icons mean? 
- What's the relationship between the bottom 4 items (crystal nova, etc) with the character? Are they spells that she can cast?
- Finally, is she "crystal maiden" or "Rylai"? Title says the former, but everywhere it seems to refer to her as the latter...

Thanks for your help!

echosa commented:
1.) Not sure about getting a larger image. I’m using the same image as the source page I pull from. It looks bigger there. Do we want a larger image? I thought space was an issue. http://www.dota2.com/hero/Crystal_Maiden/ <http://www.dota2.com/hero/Crystal_Maiden/>

2.) The six icons are: (blue) intelligence, (green) agility, (red) strength, (sword) damage, (boot) speed, (shield) armor.

3.) The bottom four items are her spells. Some characters have more than four.

4.) Her name in game is “Crystal Maiden” but in her lore, she is referred to as “Rylai” (her name from Dota 1). That text is all official text.

Also, another thing to look is is searching for a character that doesn’t have intelligence as the main attribute. For instance, Riki for agility and Axe for strength. Notice the gold circle around the blue intelligence icon in your screenshot? That indicates the main attribute. In your screenshot, it already isn’t lined up right, but that’s more obvious for agility and strength. That’s a UI fix I couldn’t quite get right. I guess my CSS skills are lacking.

> On Aug 4, 2015, at 12:48 PM, Abe Yang <notifications@github.com> wrote:
> 
> @echosa <https://github.com/echosa> , forgive me for my complete lack of knowledge of Dota2 (gasp), but this is what I'm seeing on my dev machine as of yesterday:
> 
>  <https://cloud.githubusercontent.com/assets/794084/9068033/bfaa0282-3a95-11e5-8997-0d4e9b40cea6.png>
> Few questions:
> 
> Is that the largest thumbnail available to us? Can we get a bigger one?
> What do the six icons mean?
> What's the relationship between the bottom 4 items (crystal nova, etc) with the character? Are they spells that she can cast?
> Finally, is she "crystal maiden" or "Rylai"? Title says the former, but everywhere it seems to refer to her as the latter...
> Thanks for your help!
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/duckduckgo/zeroclickinfo-spice/pull/1960#issuecomment-127690112>.
>

moollaza commented:
@echosa sorry for the silence here. Just to let you know we're discussing internally the best design/implementation for this IA. We're hoping to make a new template from whatever design we decide upon so it can be applied to other game/card based IA's. The goal being to represent as much as this important info as possible so you don't have to click through ;)

echosa commented:
Thanks for the update! I'm glad I was able to inspire a whole new template from which all DDG users may benefit. :-D Keep me updated, please! I look forward to seeing this through! ^_^

abeyang commented:
@echosa Thanks for patiently bearing with us... we've went through multiple rounds with this one. :smile: 

![dota stats v5c](https://cloud.githubusercontent.com/assets/794084/9313189/5fc2c048-44d6-11e5-9ef8-2ab78917b93a.jpg)

As this is something we haven't done before, you'll have to create custom styling for this -- but hope you're game! And if you need help, @moollaza can help you get started. Also, a few things to note:
- While the image is 100% height (pulled from the website you pointed to), the width is slightly cropped, and it's horizontally centered.
- Although Crystal Maiden only has 4 abilities, I included 8 to showcase what it would look like for one who does have eight. (LMK if some characters have more than 8... that was my highest count.)

LMK if you have any more questions!

echosa commented:
@abeyang Wow. That looks really nice. I like the large hero portrait. I have two main questions, though:

1.) The icons for the abilities don't really mean much. Is there anyway to get the ability information listed, as well? Possibly in an expandable/collapsable div?

2.) How does this new template work with items? We've been focusing so much on heroes, that I fear the item search has been overlooked. Example search: "dota 2 blink" should show the item "Blink Dagger".

abeyang commented:
@echosa oops, forgot to mention that the abilities should have an `alt` tag associated with it, so that the name will be revealed on hover. (We may do a tooltip in the near future for that.)

And your'e right, this would only be for heroes -- we'll have to use another template for the items.

echosa commented:
Concerning abilities - So the plan is to only have the ability name? No chance for additional information? (At minimum the description/lore, as in my previous screenshot.)

Concerning items - Ok, so what's the plan with that? Will that need to be a separate search in order to use a different template than the hero search? I'd think that my currently search code could simply choose the template on the fly, since it already determines if it's a hero search or an item search.

Also, is the item template something you're working on already? 

Is there anything else you need from me or that I need to do to help out and keep this moving forward? :-)

moollaza commented:
@echosa I've asked Abe to answer your questions.

> I'd think that my currently search code could simply choose the template on the fly, since it already determines if it's a hero search or an item search.

Yup that sounds fine. Your JS can check the result type and based on that switch which template it passes along to the `content` property.

>Also, is the item template something you're working on already? 

I don't *think* so...
>Is there anything else you need from me or that I need to do to help out and keep this moving forward? :-)

For now, if you wouldn't mind taking a stab at implementing this design for character searches it would be great to see how it looks and works.

You'll need to use the `Base` template group here, given that this is a fully custom solution.

echosa commented:
@moollaza Oops! I really thought I'd replied to this already. I'm slightly confused. Was there not a new custom template being made for this? I was under the impression there'd be a new template already made up that I could use, but it sounds like you want me to implement your designer's design manually. I don't mind giving it a shot, but wasn't the point of this whole ordeal to have *less* custom code? That's why I thought the end result of the designer's work was going to be a new, already written template I could use in place of `List`, `Base`, etc.

abeyang commented:
@echosa Before we turn something into a template, we usually have one or two IAs use custom code; a lot of testing and netting out design (and technical) problems happen at this level. Because this is so rich in information, we figure this would be a good start for the template -- but again, at this stage, it's gonna be custom code before we can templatize this.

We hope you're game!

Also, re: **items** -- how important are item searches compared to queries on heroes? In other words, is this something DOTA players would query often?

jagtalon commented:
Woot! Excited for this

echosa commented:
@abeyang I'm certainly game! I'll get started on the new template as soon as I can.

As for item searches, I'd think it'd be as desired as heroes. For instance, when watching tournaments, if an unknown hero *or* item is picked, being able to quickly get info on that hero or item would be useful.

I'm concerned that your asking. Is this a hint that we may drop that part of this IA? I'd really like to have the item search implemented as well, as a future user of this IA myself.

abeyang commented:
@echosa I ask because I really don't know, and we do have to assess which things are important enough to tackle. Heroes are appealing because that can be easily applied to more Instant Answers (games), but if items are important, I do think it is worth tackling a well.

Why don't you take a stab at the heroes, while I take a look at the items. However, just as a fair warning, this *may* have to be split into two releases.

echosa commented:
Update: So I got on tonight to work on this a bit, and now the source URL I'm using for the JSON data isn't working. It's not even working on the main site, as you can see if you click on any hero at http://www.dota2.com/heroes/, so it's something on their end. As long as this JavaScript feed isn't working, I can't really work on this. (I *could*, but I don't want to manually fabricate a JSON response and use that instead of the real JSON response.)

abeyang commented:
@echosa it looks like it's working now?

zekiel commented:
https://duck.co/ia/view/dota2

echosa commented:
Sorry, all. I've been super busy lately and haven't had time to touch this at all. I hope to find some time soon to try and implement the new design.

@zekiel I'm not sure what the URL is or the context for it.

moollaza commented:
>Sorry, all. I've been super busy lately and haven't had time to touch this at all. I hope to find some time soon to try and implement the new design.

@echosa it's no problem. We're in no rush :+1: 

> I'm not sure what the URL is or the context for it.

Sorry about the confusion. That's the unique URL for this Instant Answer's "Instant Answer Page". We place a link to the page in the PR so that our Community Platform knows this PR is connected to the Instant Answer

echosa commented:
@moollaza @abeyang I finally got some time to rework the display. (Hooray!) I took the time to rework heroes *and* items. Here are some screenshots, including items both with and without certain info (like notes) and a hero with *lots* of abilities (he's an edge case). Note that all images now have titles, so you get text when you mouse over (also nicer for screen readers). Also, items with recipes still make each ingredient a link to the DDG Dota2 search for that item (very convenient!).

As best as I can tell, there are three main differences between what I have made and the example mock-up I was given:

1. Hero ability images are larger in my version than in the mock-up.
2. The line spacing is different. The mock-up has more space, but I didn't take the time to add that yet. It may not be necessary. Just let me know.
3. In the mock-up, the background of the instant answer area is white. When I run test searches, it is grey. 

<img width="984" alt="screen shot 2015-11-29 at 3 09 38 pm" src="https://cloud.githubusercontent.com/assets/133786/11460107/e8688a46-96ab-11e5-99c6-273c965f5388.png">

<img width="964" alt="screen shot 2015-11-29 at 3 09 54 pm" src="https://cloud.githubusercontent.com/assets/133786/11460108/f164aa30-96ab-11e5-8db4-4110c6fcc2d3.png">

<img width="976" alt="screen shot 2015-11-29 at 3 09 03 pm" src="https://cloud.githubusercontent.com/assets/133786/11460110/f6da1b6c-96ab-11e5-99bb-cb636ef3fbc2.png">

<img width="982" alt="screen shot 2015-11-29 at 3 09 24 pm" src="https://cloud.githubusercontent.com/assets/133786/11460111/01e6645c-96ac-11e5-8acb-eb0bba4cdedb.png">

echosa commented:
Oops. Just noticed the line wrapping in the attribute numbers for Invoker. I've adjusted the padding to fix it. Here's an updated screenshot for Invoker:

<img width="783" alt="screen shot 2015-11-29 at 4 40 06 pm" src="https://cloud.githubusercontent.com/assets/133786/11460738/1771f406-96b8-11e5-8c0f-e088d1a1af51.png">

moollaza commented on share/spice/dota2/content.handlebars:
```diff
@@ -0,0 +1,149 @@
+{{#if isItem}}
+    <div style="float: left; width: 100px;">
```
Let's use an external stylesheet for all the CSS. You can create `/share/spice/dota2/dota2.css` and it will automatically be loaded for you.

You'll need to prepend all the CSS rules with `.zci--dota2` though, so that your CSS only targets the elements inside your IA's `<div>` -- we automatically create a wrapper div and give it a class with `zci--<ia_id>` for you.

----

Also, we have existing classes you can use to float left/right: `pull-right`, `pull-left`. Similarly we have classes and CSS pre-made to handle grid layouts that will work on mobile so it may be best to use those.

You can read about our other classes (font sizes, background colors, etc.) here: https://duckduckgo.com/styleguide

moollaza commented on share/spice/dota2/content.handlebars:
```diff
@@ -0,0 +1,149 @@
+            Cost
+        </div>
+        <div class="eighty">
+            <img src="http://cdn.dota2.com/apps/dota2/images/tooltips/gold.png" title="Gold"/> {{cost}}
```
You'll need to use our `imageProxy` helper here so the client's IP isn't leaked to dota2.com: https://github.com/duckduckgo/duckduckgo-template-helpers/blob/ee1713f5218c18a880a133583415cb09d8088a1b/template_helpers.js#L336

moollaza replied:
Seeing as there's so many, you could build proxied URLs in the JS and pass them along here and do something like:

`<img src="{{assets.gold.img}}" title="{{assets.gold.title}}" />`

moollaza commented on share/spice/dota2/content.handlebars:
```diff
@@ -0,0 +1,149 @@
+    <div style="clear: both;" />
+{{else}}
+    {{#if isHero}}
+        <div style="float: left; width: 210px;">
```
If you prefer, you can actually create sub-templates in the share dir and use them here (or inside other templates) `include 'template_name`

> https://github.com/duckduckgo/duckduckgo-template-helpers/blob/ee1713f5218c18a880a133583415cb09d8088a1b/template_helpers.js#L363

moollaza commented on share/spice/dota2/dota2.js:
```diff
@@ -0,0 +1,148 @@
+            $.each(item.components, function(index, value) {
+                var component = {
+                    'name': fullResult.itemdata[value].dname,
+                    'img': 'http://cdn.dota2.com/apps/dota2/images/items/' + fullResult.itemdata[value].img,
```
You'll want to use our DDG proxy function here for the images: https://github.com/duckduckgo/duckduckgo-utils/blob/f1a90299a4a7330681902e91f59dc650f791fb4c/util.js#L255

moollaza commented on t/Dota2.t:
```diff
@@ -0,0 +1,42 @@
+    # This is a sample query, just like the user will enter into the DuckDuckGo
+    # search box.
+    'dota2 blink' => test_spice(
+        # The first argument is the Spice callback. It includes the javascript
```
These comments can be removed, they're just for developers to read

moollaza commented on share/spice/dota2/dota2.js:
```diff
@@ -0,0 +1,148 @@
+
+        Spice.add({
+            id: "dota2",
+            name: "Dota2",
```
Let's change this to "Games" so it corresponds with on of our existing Instant Answer Topics

moollaza commented on share/spice/dota2/dota2.js:
```diff
@@ -0,0 +1,148 @@
+            if (typeof hero !== 'undefined') {
+                returnData = createHeroReturnObject(hero, api_result);
+            }
+            sourceUrl = 'http://www.dota2.com/hero/' + hero.dname.replace(" ", "_");
```
It looks like they support HTTPS, so we might as well send them to the secure site :+1:

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=dota2 blink)
                ![](https://ia-screenshots.s3.amazonaws.com/dota2_index.png)
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/dota2_mobile.png)

echosa commented:
@moollaza I believe I have addressed all of the concerns in your review. I've also squashed down to one commit. Please review the code again and let me know if there's anything else I need to do. I'm very excited to finally have made progress on this!

echosa commented:
Bump? Just checking in on this.

moollaza commented on share/spice/dota2/dota2.js:
```diff
@@ -0,0 +1,190 @@
+        return abilities;
+    }
+
+    function getHeroPrimaryAttribute(hero) {
```
Seems like this could be replaced with an object lookup?

e.g. `primaryAttribute: primaryAttributes[hero.pa],`

moollaza commented on share/spice/dota2/dota2.js:
```diff
@@ -0,0 +1,190 @@
+            return Spice.failed('dota2');
+        }
+
+        var searchTarget = DDG.get_query().replace("dota2", "").replace("dota 2", "").trim().toLowerCase();
```
I think we can simplify this a bit: `.replace("dota2", "").replace("dota 2", "")` -> `.replace(/dota\s?2/, "")`

moollaza commented on t/Dota2.t:
```diff
@@ -0,0 +1,25 @@
+        '/js/spice/dota2/crystal%20maiden',
+        call_type => 'include',
+        caller => 'DDG::Spice::Dota2'
+    ),
```
let's add some tests that shouldn't trigger, e.g. `'dota2' => undef`

moollaza commented:
@echosa sorry for the delay! I've set this up for re-deployment to beta

> @moollaza I believe I have addressed all of the concerns in your review. 

Excellent. Thanks!

> I've also squashed down to one commit. 

TBH, I think I'd prefer if you avoid that in the future, at least until the end of the PR. The reason being, when I look at the commit, i can' see what was changed since giving my feedback. I actually prefer seeing a series of commits that show how the PR changed over time :+1: 

I'll get this on beta shortly and then we can battle test it on different themes/browsers/devices :+1:

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=dota2 blink)
                ![](https://ia-screenshots.s3.amazonaws.com/dota2_index.png)
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/dota2_mobile.png)

moollaza commented:
@echosa looks like we have some `<br>` that I think we should be able to remove because we're using `<div>`. For the Notes, what if we use a `ul` and disable to the bullets? (i.e. `list-style-type:none`)

The `<hr>` aren't showing for me in Firefox:

![dota_2_crystal_maiden_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/11694101/33b5976a-9e76-11e5-9eb5-e6a866b0b0a0.png)

We may need to use borders + padding as a better solution (or we need to adjust the CSS for the <hr>)

moollaza commented:
@echosa looks like the mobile view needs some help.

You can use classes like `half--screen-s` or `whole--screen-s` to have different widths for smaller screens.

Also, I think to ensure the content wraps properly you'll need to use our `gw` and `g` classes on the grid wrapper and elements. Checkout https://duckduckgo.com/styleguide#grids (the "Responsive Grids" section gives some details)

echosa commented:
@moollaza I've addressed almost all of your comments. I didn't do anything with Notes, because that is just a string given to us from the backend. There is no list. I could split the string on newlines to create a list, I suppose, but I didn't think it was worth it.

Also, the gold icon is no longer showing up. I don't know why it isn't, because the source works fine:

http://cdn.dota2.com/apps/dota2/images/tooltips/gold.png'

Perhaps it's something with the DDG wrapper?

Outside of those two issues, I think this is better. I've tested it on Safari, Firefox, Chrome, and iPhone 6 Plus (portrait and landscape). 

Let me know what you think!

moollaza commented:
@echosa Apologies for the delay, and Happy New year! I will get the updated version on Beta right now for review.

Thanks for doing such a great job so far! :cake: We really appreciate you taking the time to contribute. As a thanks, we'd like to write a short blog post about you and your Instant Answer.

If you're interested please send an email to open@duckduckgo.com, with the subject "DuckDuckHack Dev Appreciation".

Please provide your name and include a link to this pull request as well.

We'll get in touch for more details!

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=dota2 blink)
                ![](https://ia-screenshots.s3.amazonaws.com/dota2_index.png)
                [Mobile screenshot](https://ia-screenshots.s3.amazonaws.com/dota2_mobile.png)

echosa commented:
@moollaza So, aside from the gold image not showing up (not sure why, as explained in my previous comment), is there any more work to be done on this PR? It sounds like we're getting close to the end of it, which is exciting!

moollaza commented:
>Also, the gold icon is no longer showing up. I don't know why it isn't, because the source works fine:

Hmm this is odd. The output of the JS function is correct also -- if you manually replace the output of the function with the image it works as expected. The problem is that for some reason the image simply has no `src`. 

It seems like all other images are working though? Is the path to the image property correct in the template?

-----

>So, aside from the gold image not showing up (not sure why, as explained in my previous comment), is there any more work to be done on this PR?

@echosa I demoed this IA for the team today and I believe @abeyang has some thing he'd like to point out.

@abeyang any comments?

moollaza commented on share/spice/dota2/dota2_item.handlebars:
```diff
@@ -0,0 +1,60 @@
+            Cost
+        </div>
+        <div class="eighty eighty--screen-s seventy--screen-xs">
+            <img src="{{assets.gold.img}}" title="{{assets.gold.title}}"/> {{cost}}
```
@echosa figured it out! This should be `gold.image` :+1:

echosa replied:
D'oh! I should have caught that. :-P I'll fix that, but I'll wait to see if there are any other changes needed.

echosa commented:
@moollaza I went ahead and fixed that image variable name. It's always the little things that slip through the cracks.

So, at this point, I've addressed everything except for @abeyang potential comments, right? I'll be out of touch at least through this weekend, so if there's any other quick changed that need to be made, I might be able to get them out before I'm gone.

moollaza commented:
@echosa I spoke with the design team and the hero template still needs some work to get it looking like the original design with the white background. The whole tile should be the height of the API image.

I'm going to fiddle around tonight to see If I can adjust your CSS to achieve that look -- shouldn't be too hard. 

The changes are: 
- use the `tx--17` and `tx--13` to set the title and other text sizes correctly
- narrow the image to make room for the detail
- smaller abilities and stats images
- white background

With regards to the items, we agreed that the design we have isn't going to work out for now. So we're going to revisit a custom template for items later. We're hoping Dota2 would have bigger images for the items, but looking at the website it looks like they're just stretching the images to make them bigger which is a little disappointing. Maybe if they see the work you've done, they'll be able to give us bigger images ;)

For now, let's simplify the items by using a built-in template. I will also play around with it but I think using the `icon` template we can achieve a very similar look.

The title will be the name, subtitle can be the cost. The notes and description as you have them can be moved into a new `items_content.handlebars` template and referenced like so:

```js
templates: {
  group: 'icon',
  options: {
    content: Spice.dota2.items_content
  }
}
```

---

I'll play around with this stuff tonight any let you know how it goes!

echosa commented:
@moollaza Any progress with the template updates? I'm curious to see how the changes affect the template, so I can learn more about how they work in general.

moollaza commented:
@echosa I didn't have a chance to finish it, but I've spoken with one of our community members, who has some design/frontend experience, and he will be tackling it. I'll let you know as soon as it's ready.

altern8tif commented:
@moollaza @echosa hey guys sorry for the delay... been a bit swamped with work recently.

moollaza commented:
@altern8tif thanks! This is looking awesome.

@echosa would you like to add @altern8tif 's code to your PR so we can test it out on Beta?

For the Items, we can switch to our built-in Icon template for now

echosa commented:
For now, I'm just looking at heroes. The item page changes can happen a little later. I've added the @altern8tif's changes to my code and pushed the commit. "Thanks for the help, by the way!" Here's what it looks like, in Safari at least:

<img width="777" alt="screen shot 2016-01-24 at 1 59 35 pm" src="https://cloud.githubusercontent.com/assets/133786/12538588/e6e35dc2-c2a2-11e5-8024-7f8e30ca8022.png">

The icons just seem a little small. That concerns me more with the ability images. I tried increasing their size to 2.5em instead of 1.85em, and I think it looks better (and is easier to see what the icons are):

<img width="794" alt="screen shot 2016-01-24 at 1 59 17 pm" src="https://cloud.githubusercontent.com/assets/133786/12538602/2fc7994a-c2a3-11e5-9d54-f30283020a8d.png">

Thoughts on that icon size? I didn't push that change.

Also, I usually use DDG with a dark theme, so I decided to try that with this code. The results are not so good:

<img width="797" alt="screen shot 2016-01-24 at 1 56 13 pm" src="https://cloud.githubusercontent.com/assets/133786/12538605/4e8783a4-c2a3-11e5-848c-55f0d2c4bc88.png">

The text is grey and difficult to read.

BTW, @altern8tif, if you'd like, I can add you to the attribution section as one of the authors. Just let me know!

altern8tif commented:
@echosa Ahh... Didn't test it with the other themes/colors. My bad.  I suspect that we may need to use the `tile` class so that the tile background and border colors will invert.  I couldn't find the `tile` class in DDG's public repos so I added in some css to the `.zci--dota2 .zci__body` classes.  Let me take another look at rectifying this.

Also, I left out the `moreAt` option because the mockup that @abeyang did didn't have that, and more importantly I couldn't really decide where is the best place to put it.  Having it right below the hero image looks a little off, especially with all that white space... Any thoughts?

For the icon sizes, maybe we can get inputs from @abeyang or @moollaza , so that we can get some kind of consistency across the other IAs as well.

No worries on attribution, I'm just happy to be able to help =)

moollaza commented:
@echosa thanks for updating! I'll push to beta :+1: 

@echosa @altern8tif regarding color and the "More at" -- I forgot to mention we should switch this to use the `tile` view, instead of the detail view. That will force it to render a tile with a white background by defualt, and will push the "More At" link to the top right corner.

We can do that pretty easily with:

```js
group: 'base',
detail: 'false',
item-detail: 'false'
```

We may also need to set the `meta.primaryText = ""` to ensure we don't show any text at the top left.

moollaza commented:
Regarding the icons, I _think_ the updated size is OK. @abeyang can have the final say though

echosa commented:
Sorry, I've been _super_ busy lately. Is this waiting on more work from me, or is the work still being passed off to more experienced folks? I've lost track of where we are. :-P

moollaza commented:
@echosa apologies for the delay, somehow I didn't get your message in my email!

We're waiting on the design UI to be switched to a tile view instead of a detail view. That will force the `more at` into the metabar at the top right and put a border around the whole tile.

We also need to move the items results into a standard info template to keep things simple for now.

Let me know if you need help with anything!

echosa commented:
Sorry. Just now getting back to work on this. I'l see what I can do about the requested changes. (The red "staff attention requested" tag is slightly frightening...)

echosa commented:
Ok, so... I'm going to need help with the items results. I couldn't get it to look right in an "info" template, because the data source I retrieve from sends back HTML, which apparently doesn't get processed but, rather, gets displayed verbatim in the info template. Halp.

As for hero results, I think I figured it out. Along with setting "detail" and "item-detail" to false, I also had to  set the tile variant to "xwide". The results, I think are quite nice:

<img width="1061" alt="screen shot 2016-04-09 at 10 49 58 am" src="https://cloud.githubusercontent.com/assets/133786/14405004/f687d25c-fe49-11e5-9c77-2c5f44aded95.png">

<img width="374" alt="screen shot 2016-04-09 at 11 55 04 am" src="https://cloud.githubusercontent.com/assets/133786/14405008/00915e44-fe4a-11e5-8e52-2b3a51c4ee75.png">

Additionally, I figured out how to get rid of the `content.handlebars` files, which simply changed which custom handlebars view to show depending on if the results are a hero or an item. I was doing this:

```
{{#if isItem}}                                                                                                                                          
    {{{include 'DDH.dota2.dota2_item'}}}                                                                                                                
{{else}}                                                                                                                                                
    {{#if isHero}}                                                                                                                                      
        {{{include 'DDH.dota2.dota2_hero'}}}                                                                                                            
    {{/if}}                                                                                                                                             
{{/if}}  
```

but it turns out I can take that `include` argument, and simply pass it as `options.content`:

```
options: {
    content: DDH.dota2.dota2_hero
}
```

So, that was a nice bit of cleanup I was able to do. Yay for learning things!

Let me know what you think of the new hero tile layout, and please help me with the item view.

moollaza commented:
Hey @echosa thanks for the awesome work!

> I'm going to need help with the items results. I couldn't get it to look right in an "info" template, because the data source I retrieve from sends back HTML

Can you clarify this -- do you mean they're sending back HTML text in the JSON? Would you prefer the HTML to be sanitized? We have helper functions (both Handlebars or JS) that can strip out HTML: 

https://github.com/duckduckgo/duckduckgo-template-helpers/blob/master/template_helpers.js#L818

You can use that in the template or just use `DDG.strip_html()` in your code to send clean text to the template.

moollaza commented on share/spice/dota2/dota2.js:
```diff
@@ -154,15 +154,34 @@
+            template = {
+                group: 'base',
+                detail: false,
+                'item-detail': false,
```
this should be `item_detaill: false`

e.g. https://github.com/duckduckgo/zeroclickinfo-spice/search?utf8=%E2%9C%93&q=item_detail

echosa replied:
Ah. OK, I can fix that.

moollaza commented on share/spice/dota2/dota2.js:
```diff
@@ -154,15 +154,34 @@
             returnData = createItemReturnObject(item, api_result);
             sourceUrl = 'https://www.dota2.com/items/';
+            template = {
+                group: 'base',
```
you'll want to specify `item: false` -- so it doesn’t try to use an item (tile) template

echosa commented:
Yes, the JSON response has HTML in it. Currently, my custom handlebars file simply displays those text blocks as is, which looks like because the formatting is already there in the given HTML. That became an issue when I tried to move some of the data into an "infobox" on the right ride, which apparently doesn't support images or HTML (or anything by straight text, from what I can tell). 

As for what I'd like to do... I'm not sure. I just want the item results to look good, work in both light and dark themes, and be DDG approved. I'm not sure how to achieve that. I tried making the item search results a tile view like I've done with heroes, but just couldn't get it to look right. I think I had the same issue of the HTML being displayed verbatim instead of processed and sent to the browser raw for the formatting it provides.

echosa commented:
Ok, I've fixed those two code changes you mentioned.

echosa commented:
Ok, I just tired something again. I made the items use the same settings as heroes. That make the dark theme work, but it still doesn't look great. Here's some screenshots. Note that, as an example of HTML being passed from the JSON, the "Attributes" are broken into three lines (in this case, anyway... some items may have more, less, or no attributes at all). Those three lines come down in the JSON as HTML with `<br>` tags. Displaying verbatim is easier than having to try and parse out the correct attributes to display.

<img width="1129" alt="screen shot 2016-04-11 at 10 37 56 am" src="https://cloud.githubusercontent.com/assets/133786/14432957/f01d677c-ffd1-11e5-8937-26e6c8bb583e.png">

<img width="458" alt="screen shot 2016-04-11 at 10 38 13 am" src="https://cloud.githubusercontent.com/assets/133786/14432963/f7034e4e-ffd1-11e5-887c-1c409f4814da.png">

moollaza commented:
@echosa let me clarify -- the goal of using the info/icon template is to simplify the implementation, ideally relying on the templates's built-in handling of image/title/subtitle. Ideally we could not using any handlebars by leveraging our `record` sub-template and/or the InfoBox.

This is the direction it should head:

![dota_2_blink_dagger_at_duckduckgo](https://cloud.githubusercontent.com/assets/873785/14433591/df7e7b6c-ffdc-11e5-8f2a-13a37bb08e02.png)

The code looks like..

```js
    function createItemReturnObject(item, result) {
        return {
            isItem: true,
            title: item.dname,
            subtitle: "Cost: " + item.cost,
            description: item.desc,
            image: DDG.getImageProxyURL('http://cdn.dota2.com/apps/dota2/images/items/' + item.img),
            notes: item.notes,
            attributes: item.attrib,
            components: getItemComponents(item, result),
            assets: getItemDisplayAssets()
        };
    }
```

```js
        if (typeof item !== 'undefined') {
            returnData = createItemReturnObject(item, api_result);
            console.log(returnData);
            sourceUrl = 'https://www.dota2.com/items/';
            template = {
                group: 'info',
                item: false,
                options: {
                    content: Spice.dota2.dota2_item,
                    moreAt: true
                }
            };
```

```handlebars
{{#if description}}
    <div class="text--secondary">
        Description
    </div>
    <div>
        {{{description}}}
    </div>
{{/if}}
{{#if notes}}
    <div class="text--secondary">
        Notes
    </div>
    <div>
        {{{notes}}}
    </div>
{{/if}}
{{#if attributes}}
    <div class="text--secondary">
        Attributes
    </div>
    <div>
        {{{attributes}}}
    </div>
{{/if}}
{{#if components}}
    <div class="text--secondary">
        Recipe
    </div>
    <div>
        <table>
            <tr>
                {{#each components}}
                    <td class="item-component">
                        <a href="/?q=dota2+{{searchTerms}}&ia=dota2">
                            <img src="{{img}}" title="{{name}}"/>
                        </a>
                    </td>
                {{/each}}
            </tr>
        </table>
    </div>
{{/if}}
```


A good use of the infobox can be seen in the Pokemon Fathead: https://duckduckgo.com/?q=wartortle&ia=about

If we can move key-value out into the InfoBox that'd be good.

Does this make sense?

moollaza commented:
You'll also want to remove the first CSS rule that modifies the whole `.zci` -- it shouldn't be needed if we're using the templates

echosa commented:
> A good use of the infobox can be seen in the Pokemon Fathead

Yeah, I looked at that as an example when I tried to implement an infobox before. The main issue was the HTML coming down from the JSON. I guess I'll need to manually parse that out, and hope they don't change the structure anytime soon. I can't just strip the HTML tags, because I need to get an array or list of the items it displays. I'll see what I can do about that. Having to manually parse the HTML was a last resort, to be sure, but it sounds like that's where I find myself.

echosa commented:
@moollaza Ok, so, here we. I moved the item results into a basic info view, with one exception: the components. Having clickable images that search for that item was too nice, so I moved that into a sub-template, and removed the old custom items template. I couldn't get the HTML to parse quite how I wanted, but simply stripping the HTML at least got the info into the info box. Here's a few screenshots of various items, some of which have attributes, notes components, etc., some of which do not.

<img width="1152" alt="screen shot 2016-04-11 at 12 33 34 pm" src="https://cloud.githubusercontent.com/assets/133786/14436510/a66beebc-ffe2-11e5-85d7-a24e40d818b5.png">
<img width="1153" alt="screen shot 2016-04-11 at 12 33 22 pm" src="https://cloud.githubusercontent.com/assets/133786/14436511/a68345e4-ffe2-11e5-9691-d1bb7f7efea7.png">
<img width="1145" alt="screen shot 2016-04-11 at 12 32 09 pm" src="https://cloud.githubusercontent.com/assets/133786/14436512/a685433a-ffe2-11e5-860b-d15aed368669.png">

moollaza commented:
@echosa thi is almost there -- let's bring the description back to the main body -- it's too lengthy for the InfoBox, especially in the Blink Dagger photo.

Is there ever a chance of having a description, notes, _and_ components? If so, can we ignore the notes in those cases?

Long text should be in the body, key-value info should be in the infobox.

Let's reduce the image size on the components and maybe put a little space between them?

echosa commented:
@moollaza 

> Is there ever a chance of having a description, notes, and components?

I'll do you one better: there's a chance of having all four: description, notes, attributes, *and* components. 

>  If so, can we ignore the notes in those cases?

We can probably ignore the the notes in all cases. It's good info, sure, but maybe a little too complete for the DDG IA. I think I'm just trying to stuff to much info in at once, but if we have to cut something out, I think the notes are a good candidate. Description and Attributes are important, though, I think. Which brings me to...

> Long text should be in the body, key-value info should be in the infobox.

So, given the screenshots above, can I leave description and attributes in the infobox (and just remove notes entirely)? If not, we're back to what I was doing before: putting that information into the custom content handlebars file.

> Let's reduce the image size on the components and maybe put a little space between them?

I can do that. I was thinking the same already, actually.

moollaza commented:
> So, given the screenshots above, can I leave description and attributes in the infobox (and just remove notes entirely)? If not, we're back to what I was doing before: putting that information into the custom content handlebars file.

Let's put the description back in the body -- that's totally fine and that's where we expect it to be.

Attributes can stay in the InfoBox. If you can break them up so they're line-by-line, instead of a paragraph that'd look awesome

echosa commented:
Ok, I'll put description back in the body. Does that mean I can put the notes in there, too? :-D

As for breaking up the attributes, I'll try again, but I was unable to do so last time. Couldn't get the string split or regex match quite right. Plus, I'm not even sure if all items' attributes will be formatted the same way, so I'm can't be certain a regex would work unless I tree all the items. I suppose I can take a sample set and try it, though.

echosa commented:
As I'm working on this, I found that the documentation about infobox nested properties here seems wrong:

http://docs.duckduckhack.com/frontend-reference/templates-reference.html

When I try that example with the attributes, this is what I get:

<img width="535" alt="screen shot 2016-04-14 at 6 24 16 pm" src="https://cloud.githubusercontent.com/assets/133786/14546893/21019d92-026e-11e6-99a6-34280346df18.png">

echosa commented:
Ok, I just pushed changes to the item search results. Here's some screenshots:

<img width="966" alt="screen shot 2016-04-14 at 6 29 28 pm" src="https://cloud.githubusercontent.com/assets/133786/14546999/5100b2de-026f-11e6-8d7c-d79340ec6a3c.png">

<img width="955" alt="screen shot 2016-04-14 at 6 29 14 pm" src="https://cloud.githubusercontent.com/assets/133786/14547001/54fd2d04-026f-11e6-8095-4d261993d5ca.png">

<img width="1039" alt="screen shot 2016-04-14 at 6 27 34 pm" src="https://cloud.githubusercontent.com/assets/133786/14547005/5a7cdb58-026f-11e6-97b7-0d574876a548.png">

One issue is when I size down to mobile size, the infobox disappears:

<img width="567" alt="screen shot 2016-04-14 at 6 28 14 pm" src="https://cloud.githubusercontent.com/assets/133786/14547044/9b127664-026f-11e6-847c-71ed40c7a7aa.png">

That seems to be a DDG issue, since searching DDG for "bulbasaur" results in the same behavior. Resizing down makes the infobox disappear. Sometimes the info shows on refresh, sometimes not, it seems. Either way, it's not an issue with my particular IA. Just wanted to mention it.

echosa commented:
@moollaza Given my last two comments, please take a look again.

daxtheduck commented:
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=crystal maiden dota2)
                ![](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/dota2_index.png)
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/dota2_mobile.png)

moollaza commented:
@echosa FYI it's now live on beta^

moollaza commented:
@echosa Thanks! This LGTM :+1:

There are a few minor things to fix (I'll make some comments), but this is definitely more than acceptable!

Thanks again for all the awesome work, I'm sorry it took so long, but I'm really happy with the final result!

moollaza commented on share/spice/dota2/dota2.css:
```diff
@@ -0,0 +1,49 @@
+
+.zci--dota2 .hero-image img {
+    height: 100%;
+    object-fit: cover;
```
This will break in IE. We need an approach that will work for all major browsers, or at least fallback CSS for IE 9+

moollaza commented on share/spice/dota2/dota2_hero.handlebars:
```diff
@@ -0,0 +1,71 @@
+<div class="gw">
+    <div class="g third hero-image">
+        <img class="whole--screen-s" src="{{image}}" title="{{name}}" />
```
I think we'll need to shrink the height of the image a bit here, right now the whole IA doesn't fit within the first page.

Alternatively we need to use our content chomping classes/JS to limit the height of the initial result and allow the user to expand it. If they want to see more info

moollaza commented on share/spice/dota2/dota2.js:
```diff
@@ -0,0 +1,226 @@
+
+        if (item.attrib) {
+            infoboxData.push({
+                label: 'Attributes',
```
This should be a `heading` not a `label` + `value`:

```js
{ heading: "Attributes:" },
```

moollaza commented on lib/DDG/Spice/Dota2.pm:
```diff
@@ -0,0 +1,26 @@
+
+use DDG::Spice;
+
+name "Dota 2 Hero and Item Search";
```
This metadata should be removed -- that all lives in https://duck.co/ia/view/dota2

echosa commented:
@moollaza Awesome! I can't wait to use this on the live site! :-) Do I need to address your comments since this is already merged?

moollaza commented:
@echosa apologies for the delay, I've been on vacation! This is now live: https://duckduckgo.com/?q=dota+2+blink+dagger&ia=games

Thanks again for the awesome work!

edgesince84 commented:
@echosa Hi Brian!  we really appreciate your work -- would you email me at merch @ duckduckgo .com?  we'd love to get you some swag 👍
